File: src/entity/idiom.ts
export interface CharData {
  c: string;
  py: string;
  sm: string;
  ym: string;
  tone: number;
}

export interface IdiomEntry {
  id?: number;
  word: string;
  pinyin: string;
  explanation: string;
  char_count?: number;
  chars: CharData[];
  blind_eligible?: boolean;
}

export interface IdiomInput {
  word: string;
  pinyin: string;
  explanation?: string;
}

export interface IndexedIdiom extends IdiomEntry {
  chars: CharData[];
}

export type MarkState = 'hit' | 'present' | 'absent' | null;

export interface PositionMarks {
  char: MarkState;
  sm: MarkState;
  ym: MarkState;
  tone: MarkState;
  /** 声/韵/调是否由「字」的存在标记自动带出（人选可覆盖，冲突以人选为准） */
  autoSm?: boolean;
  autoYm?: boolean;
  autoTone?: boolean;
}

export interface DeduceGuess {
  word: string;
  pinyin: string;
  chars: CharData[];
  marks: PositionMarks[];
  editIndex?: number;
}

export interface LexiconMeta {
  total: number;
  blind_count: number;
}

export interface UpsertResult {
  added: number;
  updated: number;
  skipped: number;
  errors: string[];
}

export interface BuiltinLexiconInfo {
  id: string;
  name: string;
  description: string;
}

File: src-tauri/src/entity/idiom.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CharData {
    pub c: String,
    pub py: String,
    pub sm: String,
    pub ym: String,
    pub tone: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Idiom {
    pub id: Option<i64>,
    pub word: String,
    pub pinyin: String,
    pub explanation: String,
    pub char_count: i32,
    pub chars: Vec<CharData>,
    pub blind_eligible: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IdiomInput {
    pub word: String,
    pub pinyin: String,
    pub explanation: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LexiconMeta {
    pub total: i64,
    pub blind_count: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpsertResult {
    pub added: usize,
    pub updated: usize,
    pub skipped: usize,
    pub errors: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BuiltinLexiconInfo {
    pub id: String,
    pub name: String,
    pub description: String,
}
